File: c++/day16/part1.cpp
#include "part1.h"

using namespace std;


namespace Day16 {

    /*
     * Valves with rate = 0 are useless, they are just tunnels.
     * We can remove them and connect directly the valves linked to them
     * this reduces the total number of valves of the graph
     * We keep only valve AA with a zero rate, as it is the starting point.
     */
    void reduceValves(map<string, Valve>& valves) {
        set<string> deletedValves;
        for (const auto& valveName : valves | views::keys) {
            if (valves[valveName].rate == 0 && valveName != "AA") {
                for (const auto& v1 : valves[valveName].tunnels | views::keys) {
                    for (const auto& v2 : valves[valveName].tunnels | views::keys) {
                        if (v1 == v2) continue;
                        const int distance = valves[valveName].tunnels[v1] + valves[valveName].tunnels[v2];
                        valves[v1].tunnels.insert({ v2, distance });
                        valves[v2].tunnels.insert({ v1, distance });
                        valves[v1].tunnels.erase(valveName);
                        valves[v2].tunnels.erase(valveName);
                    }
                }
                deletedValves.insert(valveName);
            }
        }
        for (const auto& valveName : deletedValves) valves.erase(valveName);
    }

    /*
     * We need to go to one valve to another, and find the best order to maximize the total rate.
     * To simplify the traversal, we enrich the tunnels to calculate the optimal path from each valve to each valve.
     */
    void expandValves(map<string, Valve>& valves) {
        for (const auto& origin : valves | views::keys) {
            map<string, int> bestDist {};
            for (const auto& dest : valves | views::keys) {
                bestDist[dest] = -1;
            }
            bestDist[origin] = 0;
            deque<string> toCheck { origin };
            while (toCheck.size() > 0) {
                const auto& curr = toCheck.front();
                for (const auto& [v, d] : valves[curr].tunnels) {
                    if (bestDist[v] == -1 || bestDist[curr] + d < bestDist[v]) {
                        bestDist[v] = bestDist[curr] + d;
                        toCheck.push_back(v);
                    }
                }
                toCheck.pop_front();
            }
            bestDist.erase(origin);
            valves[origin].tunnels = bestDist;
        }
    }

    map<string, Valve> parse(const string& filename) {
        map<string, Valve> valves;
        const regex pattern { "Valve ([A-Z]{2}) has flow rate=(\\d+); tunnels? leads? to valves? ([A-Z, ]+)" };
        for (const auto& line : getFileLines(filename)) {
            smatch match;
            regex_search(line.begin(), line.end(), match, pattern);
            Valve valve { match[1], stoi(match[2]) };
            for (const auto& destValve : split(match[3], ", ")) {
                valve.tunnels.insert({ destValve, 1 });
            }
            valves.insert({ valve.name,  valve });
        }
        reduceValves(valves);
        expandValves(valves);
        return valves;
    }

    namespace Part1 {

        struct State {
            set<string> openValves {};
            string position;
            int time;
            int score;
        };

        int solve(const map<string, Valve>& valves) {
            int bestScore = 0;
            deque<State> toCheck {};
            toCheck.push_back(State{ { "AA" }, "AA", 0, 0 });
            while (toCheck.size() > 0) {
                const auto& state = toCheck.front();
                for (const auto& nextValve : valves | views::keys) {
                    if (nextValve == "AA" || state.openValves.find(nextValve) != state.openValves.end()) {
                        // cannot open the valve there
                        continue;
                    }
                    const int nextTime = state.time + valves.at(state.position).tunnels.at(nextValve) + 1;
                    if (nextTime >= 30) {
                        // not enough time to go open that valve
                        continue;
                    }
                    State nextState { state };
                    nextState.openValves.insert(nextValve);
                    nextState.position = nextValve;
                    nextState.time = nextTime;
                    nextState.score += (30 - nextTime) * valves.at(nextValve).rate;
                    bestScore = max(bestScore, nextState.score);
                    toCheck.push_back(nextState);
                }
                toCheck.pop_front();
            }
            return bestScore;
        }

    }
}

File: python/day16/part1.py
#!/usr/bin/env python3
import collections
import re
import time


class Valve:
    def __init__(self, rate, lead_to):
        self.rate = rate
        self.lead_to = {valve: 1 for valve in lead_to}


class State:
    def __init__(self, location, valves, visited, time, score):
        self.location = location
        self.valves = valves
        self.visited = visited
        self.time = time
        self.score = score


def simplify_valves(valves):
    # many valves in the input file have a rate of zero, so they cannot be opened.
    # they can be treated as a simple tunnel between actual valves, so we remove them and simply
    # increase the distance between the connected valves
    for valve_name in set(name for name in valves):
        if valve_name == 'AA':  # starting valve, can't simplify it
            continue
        if valves[valve_name].rate == 0:
            for v1 in valves[valve_name].lead_to:
                for v2 in valves[valve_name].lead_to:
                    if v1 != v2:
                        valves[v1].lead_to[v2] = valves[valve_name].lead_to[v1] + valves[valve_name].lead_to[v2]
                        del valves[v1].lead_to[valve_name]
            del valves[valve_name]


def parse(input_path):
    with open(input_path, 'r') as f:
        valves = {}
        for line in f.readlines():
            name, rate, lead_to = re.match(
                r'Valve ([A-Z]+) has flow rate=(\d+); tunnel[s]* lead[s]* to valve[s]* ([A-Z, ]+)', line).groups()
            valves[name] = Valve(int(rate), lead_to.split(', '))
        simplify_valves(valves)
        return valves


def solve_part1(valves):
    max_time = 30
    seen = {name: {} for name in valves}
    best_score = 0
    to_check = collections.deque()
    to_check.append(State('AA', set(), {'AA'}, 0, 0))

    while len(to_check):
        curr_state = to_check.popleft()
        best_score = max(best_score, curr_state.score)

        # no more processing if all valves are opens (except AA that has rate = 0)
        if  len(curr_state.valves) == len(valves)-1:
            continue

        if curr_state.time >= max_time:
            continue

        # discard the state if no more chance to beat the current best score
        remaining_rate = sum([valves[v].rate for v in valves if v not in curr_state.valves])
        if curr_state.score + remaining_rate * (max_time - curr_state.time) <= best_score:
            continue

        # discard the state if a better state was already processed
        tuple_key = tuple(curr_state.valves)
        if tuple_key not in seen[curr_state.location]:
            seen[curr_state.location][tuple_key] = []
        if any([t <= curr_state.time and s >= curr_state.score for (t, s) in seen[curr_state.location][tuple_key]]):
            continue
        else:
            seen[curr_state.location][tuple_key].append((curr_state.time, curr_state.score))

        # open the valve
        if valves[curr_state.location].rate > 0 and curr_state.location not in curr_state.valves:
            to_check.append(State(curr_state.location,
                                  curr_state.valves | {curr_state.location},
                                  {curr_state.location},
                                  curr_state.time + 1,
                                  curr_state.score + valves[curr_state.location].rate * (max_time - curr_state.time - 1)))

        # discard the state if we have already been here since the last valve we opened
        for next_valve in valves[curr_state.location].lead_to:
            if next_valve in curr_state.visited:
                continue

            to_check.append(State(next_valve,
                                  curr_state.valves,
                                  curr_state.visited | {next_valve},
                                  curr_state.time + valves[curr_state.location].lead_to[next_valve],
                                  curr_state.score))
    return best_score


if __name__ == "__main__":
    start = time.time()
    parsed = parse("data.txt")
    result = solve_part1(parsed)
    print(f'Part 1 : {result}  ({round(1000 * (time.time() - start))} ms)')
